include/proxy-wasm/context_interface.h

```diff
@@ -0,0 +1,663 @@
+};
+
+/**
+ * SharedDataInterface is for shaing data between VMs. In general the VMs may be on different
+ * threads.  Keys can have any format, but good practice would use reverse DNS and namespacing
+ * prefixes to avoid conflicts.
+ */
+struct SharedDataInterface {
+  virtual ~SharedDataInterface() = default;
+  /**
+   * Get proxy-wide key-value data shared between VMs.
+   * @param key is a proxy-wide key mapping to the shared data value.
+   * @param cas is a number which will be incremented when a data value has been changed.
+   * @param data is a location to store the returned stored 'value' and the corresponding 'cas'
+   * compare-and-swap value which can be used with setSharedData for safe concurrent updates.
+   */
+  virtual WasmResult
+  getSharedData(string_view key, std::pair<std::string /* value */, uint32_t /* cas */> *data) = 0;
+
+  /**
+   * Set a key-value data shared between VMs.
+   * @param key is a proxy-wide key mapping to the shared data value.
+   * @param cas is a compare-and-swap value. If it is zero it is ignored, otherwise it must match
+   * the cas associated with the value.
+   * @param data is a location to store the returned value.
+   */
+  virtual WasmResult setSharedData(string_view key, string_view value, uint32_t cas) = 0;
+}; // namespace proxy_wasm
+
+struct SharedQueueInterface {
+  virtual ~SharedQueueInterface() = default;
+  /**
+   * Register a proxy-wide queue.
+   * @param queue_name is a name for the queue. The queue_name is combined with the vm_id (if any)
+   * to make a unique identifier for the queue.
+   * @param token_ptr a location to store a token corresponding to the queue.
+   */
+  virtual WasmResult registerSharedQueue(string_view queue_name,
+                                         SharedQueueDequeueToken *token_ptr) = 0;
+
+  /**
+   * Get the token for a queue.
+   * @param vm_id is the vm_id of the Plugin of the Root Context which registered the queue.
+   * @param queue_name is a name for the queue. The queue_name is combined with the vm_id (if any)
+   * to make a unique identifier for the queue.
+   * @param token_ptr a location to store a token corresponding to the queue.
+   */
+  virtual WasmResult lookupSharedQueue(string_view vm_id, string_view queue_name,
+                                       SharedQueueEnqueueToken *token_ptr) = 0;
+
+  /**
+   * Dequeue a message from a shared queue.
+   * @param token is a token returned by registerSharedQueue();
+   * @param data_ptr is a location to store the data dequeued.
+   */
+  virtual WasmResult dequeueSharedQueue(SharedQueueDequeueToken token, std::string *data_ptr) = 0;
+
+  /**
+   * Enqueue a message on a shared queue.
+   * @param token is a token returned by resolveSharedQueue();
+   * @param data is the data to be queued.
+   */
+  virtual WasmResult enqueueSharedQueue(SharedQueueEnqueueToken token, string_view data) = 0;
+};
+
+struct ContextInterface : RootInterface,
```

Comment:
You've separated this nicely into interfaces, but then you never use them separately. Shouldn't `ContextBase` inherit from all of them directly, and not via this catch-all proxy? 

Also, it seems that `ContextBase` features (i.e. `onCreate`, `onDone`, `onLog` and `onDelete`) are intermixed with everything else.

Reply:
It could.  I didn't want to expose all this up to the next level.  This was all based on suggestions by Matt.  I suppose that ContextBase could be used for mocks. but nobody seems to do that AFAICT.  If you want I can remove ContextInterface.

Reply:
Up to you. You can leave it for now.

Reply:
+1 to remove ContextInterface but I don't feel strongly either way. The main thing that I care about is the interface split above which makes it 1000x easier to understand for a new person coming into all of this.

Reply:
I'll follow up with that change.  I want to get this in so I can update the dependent envoy-wasm PR.  Since removing ContextInterface will not be visible above this repo it can be done independently.